src/MongoDB.Driver/Encryption/LibMongoCryptController.cs

```diff
@@ -0,0 +1,676 @@
+﻿/* Copyright 2019-present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Security;
+using System.Net.Sockets;
+using System.Security.Cryptography.X509Certificates;
+using System.Threading;
+using System.Threading.Tasks;
+using MongoDB.Bson;
+using MongoDB.Bson.IO;
+using MongoDB.Bson.Serialization.Serializers;
+using MongoDB.Driver.Core.Clusters;
+using MongoDB.Driver.Core.Misc;
+using MongoDB.Driver.Core.WireProtocol;
+using MongoDB.Libmongocrypt;
+
+namespace MongoDB.Driver.Encryption
+{
+    internal sealed class LibMongoCryptController : IBinaryDocumentFieldDecryptor, IBinaryCommandFieldEncryptor, IDisposable
+    {
+        #region static
+        private static bool AcceptAnyCertificate(
+            object sender,
+            X509Certificate certificate,
+            X509Chain chain,
+            SslPolicyErrors sslPolicyErrors)
+        {
+            return true;
+        }
+        #endregion
+
+        // private fields
+        private readonly MongoClient _client;
+        private CryptClient _cryptClient;
+        private bool _disposed;
+        private readonly EncryptionMode _encryptionMode;
+        private readonly IReadOnlyDictionary<string, object> _extraOptions;
+        private bool _initialized = false;
+        private readonly IMongoClient _keyVaultClient;
+        private IMongoCollection<BsonDocument> _keyVaultCollection;
+        private readonly CollectionNamespace _keyVaultNamespace;
+        private readonly IReadOnlyDictionary<string, IReadOnlyDictionary<string, object>> _kmsProviders;
+        private MongoClient _mongocryptdClient;
+        private readonly IReadOnlyDictionary<string, BsonDocument> _schemaMap;
+
+        // constructors
+        public LibMongoCryptController(
+            MongoClient client,
+            ClientEncryptionOptions clientEncryptionOptions)
+        {
+            _client = Ensure.IsNotNull(client, nameof(client));
+            _encryptionMode = EncryptionMode.ClientEncryption;
+            _keyVaultClient = Ensure.IsNotNull(clientEncryptionOptions.KeyVaultClient, "keyVaultClient");
```

Comment:
Nit: I believe this could be:
`_keyVaultClient = Ensure.IsNotNull(clientEncryptionOptions.KeyVaultClient, nameof(ClientEncryptionOptions.KeyVaultClient));`
And ditto for the rest of the `Ensure` statements